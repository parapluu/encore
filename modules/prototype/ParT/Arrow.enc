#+literate

* Arrow Module

Version 0.1
Authors:
- Kiko Fernandez-Reyes <kiko.fernandez@it.uu.se>

#+BEGIN_SRC encore
module Arrow(
    arr,
    first,
    second,
    composition,
    branch,
    combine,
    liftA2,
    runArrow,
    runParArrow
  )

import ParT
#+END_SRC

** arr :: Arrow a => (b -> c) -> a b c

Lifting a function to a ParT arrow

`arr` lifts a function to the ParT type. The ParT type represents
an abstraction with possibly many functions. By lifting multiple
functions these will potentially start performing work in parallel
as soon as the `runArrow` method is called.

#+BEGIN_SRC encore
fun arr[a, b](fn : a -> b) : Par[a -> b]
  liftv(fn)
end
#+END_SRC

** first :: Arrow a => a b c -> a (b, b) (c, d)

Given a ParT function that accepts one argument of type `a`
and returns an argument of type `b`, return a lifted function
that takes a tuple `(a, c)` and returns a tuple `(b, c)`.

As it can be observed, the second argument in the tuple is not changed.

#+BEGIN_SRC encore
--
fun first[a, b, c](p : Par[a -> b]) : Par[((a, c)) -> (b, c)]
  val closure = fun (fn : a -> b)
                  liftv(fun (t : (a, c)) => (fn(t.0), t.1))
                end
  bind(closure, p)
end
#+END_SRC

** second :: Arrow a => a b c -> A (d, b) (d, c)

Given a ParT function that accepts one argument of type `a`
and returns an argument of type `b`, return a lifted function
that takes a tuple `(c, a)` and returns a tuple `(b, c)`.

As it can be observed, the first argument in the tuple is not changed.

#+BEGIN_SRC encore

fun second[a, b, c](p : Par[a -> b]) : Par[((c, a)) -> (c, b)]
  val closure = fun (fn : a -> b)
                  liftv(fun (t : (c, a)) => (t.0, fn(t.1)))
                end
  bind(closure, p)
end
#+END_SRC


** (>>>) :: Control.Category.Category cat => cat a b -> cat b c -> cat a c

Same as function composition except that its ParT (arrow) composition

#+BEGIN_SRC encore
-- private function
fun compose[a, b, c](f1 : a -> b, f2 : b -> c) : a -> c
  fun (x : a) => f2(f1(x))
end

fun composition[a, d, c](p : Par[a -> d], p2 : Par[d -> c]) : Par[a -> c]
  bind(fun (fn : a -> d)
         bind(fun (fn2 : d -> c) => liftv(compose(fn, fn2)), p2)
       end, p)
end
#+END_SRC


** (&&&) :: Arrow a => a b c -> a b c' -> a b (c, c')

Given two ParT with the same first argument type in common,
return a new ParT that performs the execution of both functions,
yielding a tuple `(b, c)`

#+BEGIN_SRC encore
fun branch[a, b, c](p : Par[a -> b], p2 : Par[a -> c]) : Par[a -> (b, c)]
  val clos = fun (fn1 : a -> b)
               val clos2 = fun (fn2 : a -> c)
                             liftv(fun (x : a) => (fn1(x), fn2(x)))
                           end
               bind(clos2, p2)
             end
  bind(clos, p)
end
#+END_SRC

** (***) :: Arrow a => a b c -> a d e -> a (b, d) (c, e)

Given two ParTs, return a lifted ParT that takes a tuple `(a, c)`
and returns tuples `(b, d)`.

#+BEGIN_SRC encore
fun combine[a, b, c, d](p1 : Par[a -> b], p2 : Par[c -> d]) : Par[((a, c)) -> (b, d)]
  val clos = fun (fn1 : a -> b)
               val clos2 = fun (fn2 : c -> d)
                             liftv(fun (t : (a, c)) => (fn1(t.0), fn2(t.1)))
                           end
               bind(clos2, p2)
             end
  bind(clos, p1)
end
#+END_SRC

** runArrow

This function is not needed in Haskell but it is needed in the ParT
abstraction. This is because a ParT may contain possibly many
functions. When we pass an initial item we cannot know if the ParT
contains a single lifted function or many of them inside. For this
reason and to continue executing functions asynchronously, we return
a new ParT.

Utility function to lift the item into the arrow and perform function application

#+BEGIN_SRC encore
fun runArrow[a, b](item : a, p : Par[a -> b]) : Par[b]
  p >> (fun (fn : a -> b) => fn(item))
end
#+END_SRC

** liftA2 :: Applicative f => (a -> b -> c) -> f a -> f b -> f c

A bit similar in functionality to the `liftA2` from Haskell except that
we don't have support for typeclasses and the logic has been hard coded
especifically for the ParT abstraction. It resembles a ForkJoin in the sense
that two functions may run possibly in parallel and their end result
will be aggregated using the function `fn`.

#+BEGIN_SRC encore
fun liftA2[a, b, c, d](fn: (b, c) -> d, p1: Par[a -> b], p2: Par[a -> c]): Par [a -> d]
  val pb = branch(p1, p2)
  val fnArr = liftv(fun (t: (b, c)) => fn(t.0, t.1))
  composition(pb, fnArr)
end
#+END_SRC

** runParArrow

Given a ParT abstraction and a ParT arrow, lift each item in the ParT
to the arrow function and run.

#+BEGIN_SRC encore
fun runParArrow[a, b](p: Par[a], arrow: Par[a -> b]): Par[b]
  join(p >> fun (item: a) => runArrow(item, arrow))
end
#+END_SRC
