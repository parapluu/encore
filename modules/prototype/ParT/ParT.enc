module ParT(
    empty,
    filter,
    bind,
    foreachp,
    aggregate,
    intersection,
    distinct,
    union,
    app,
    pairWith,
    groupJoin,
  )

fun empty[a]() : Par[a]
  EMBED (Par[a])
    new_par_empty(_ctx, runtimeType[0]);
  END
end

fun filter[a](fn : a -> bool, ps : Par[a]) : Par[a]
  join(ps >> fun (item : a) => if fn(item) then
                                 liftv(item)
                               else
                                 empty[a]()
                               end)
end

fun bind[a, b](fn : a -> Par[b], p : Par[a]) : Par[b]
  join(p >> fn)
end

fun foreachp[t](p : Par[t], fn : t -> void) : Par[void]
  p >> fn
end

fun aggregate[t, a, r](p : Par[t], var init : a, fn1 : (a, t) -> a, fn2 : a -> r) : Par[r]
  liftf(reduce(fn1, init, p)) >> fn2
end

fun intersection[t](p1 : Par[t], p2 : Par[t], cmp : (t, t) -> int) : Par[t]
  EMBED (Par[t])
    party_intersection(_ctx, #{p1}, #{p2}, #{cmp}, _enc__type_t);
  END
end

fun distinct[t](p : Par[t], cmp : (t, t) -> int) : Par[t]
  EMBED (Par[t])
    party_distinct(_ctx, #{p}, #{cmp}, _enc__type_t);
  END
end

fun union[t](pl : Par[t], pr : Par[t], cmp : (t, t) -> int) : Par[t]
  distinct(pl ||| pr, cmp)
end

fun app[a, b](p : Par[a -> b], item : a) : Par[b]
  p >> (fun (fn : a -> b) => fn(item))
end

fun pairWith[t, k](p : Par[t], fn : t -> k) : Par[(k, t)]
  p >> (fun (x : t) => (fn(x), x))
end

-- private function.
-- filter out items in the `p` ParT whose key is different from the given key `x`.
-- return a future with the elements that match the key
fun groupParTuples[k, i](x : k, p : Par[(k, i)], cmp : (k, k) -> int) : Fut[(k, Par[i])]
  val closure = fun (acc : (k, Par[i]), elem : (k, i))
                  val accKey = acc.0
                  val p = acc.1
                  val elemKey = elem.0
                  val elemValue = elem.1
                  val equals = cmp(accKey, elemKey)
                  if equals == 0 then
                    (accKey, p ||| liftv(elemValue))
                  else
                    acc
                  end
                end
  reduce(closure, (x, empty[i]()), p)
end

-- private function
-- finds if an element exists in a ParT of tuples
fun findElemInParTuple[k, t](key : k, p : Par[(k, t)], cmp : (k, k) -> int) : Fut[Maybe[t]]
  val closure = fun (acc : Maybe[t], elem : (k, t))
                  val keyElem = elem.0
                  val keyValue = elem.1
                  if cmp(key, keyElem) == 0 then
                     Just(keyValue)
                  else
                     acc
                  end
                end
  reduce(closure, Nothing, p)
end



{-
Group elements of two ParTs based on their keys and joins them according
to their join function. Elements that have no matching key are ignored.

E.g.

```
  val magnus = new Person("Magnus")
  val terry = new Person("Terry")
  val charlotte = new Person("Charlotte")
  val kiko = new Person("Kiko")

  val barley = new Pet("Barley", terry)
  val boots = new Pet("Boots", terry)
  val whiskers = new Pet("Whiskers", charlotte)
  val daisy = new Pet("Daisy", magnus)

  val people = liftv(magnus) ||| liftv(terry) ||| liftv(charlotte) ||| liftv(kiko)
  val pets = liftv(barley) ||| liftv(boots) ||| liftv(whiskers) ||| liftv(daisy)
  val query = groupJoin(people,
                        pets,
                        fun (p : Person) => p.name,
                        fun (p : Pet) => p.owner.name,
                        fun (per : Person, animals : Par[Pet])
                          liftv((per.name, animals >> (fun (p : Pet) => p.name)))
                        end,
                        fun (s1 : String, s2 : String) => s1.compare(s2))

  for elem <- extract(query) do
    match elem with
      case (name, pets) =>
        println("{}:", name)
        for pet <- extract(pets) do
          println("{}", pet)
        end
      end
    end
  end
```

In the example above, the expected output is:

```
  Terry:
  Barley
  Boots
  Charlotte:
  Whiskers
  Magnus:
  Daisy
```

where the person `Kiko` doesn't have any pet and is therefore ignored.

-}

fun groupJoin[t, i, k, r](pt : Par[t], pi : Par[i], tk : t -> k, ik : i -> k,
                          fn : (t, Par[i]) -> Par[r], cmp : (k, k) -> int) : Par[r]
  -- Pair function with their corresponding ParT
  val pkt = pairWith(pt, tk)
  val pki = pairWith(pi, ik)

  -- Get distinct elements
  val pkid = distinct(pki, fun (v : (k, i), u : (k, i))
                             cmp(v.0, u.0)
                           end)

  join(pkid >> (fun (tuple : (k, i))
                  val key = tuple.0
                  -- group values from the same key under the same ParT
                  val groupedPiKeyToPar = groupParTuples(key, pki, cmp)

                  -- fetch only the values from the tuple
                  val groupedPi = join(liftf(groupedPiKeyToPar ~~> (fun (keyParT: (k, Par[i]))
                                                                      keyParT.1
                                                                    end)))

                  -- find if the key from second ParT (from pkdi) exists
                  -- in the pkt ParT and execute function asynchronously
                  -- only if it does exist.
                  val futtkElem = findElemInParTuple(key, pkt, cmp)
                  val fp = futtkElem ~~> fun (tkElem : Maybe[t])
                                           match tkElem with
                                             case Nothing =>
                                               empty[r]()
                                             end
                                             case Just(tkResultVal) =>
                                               fn(tkResultVal, groupedPi)
                                             end
                                           end
                                         end

                  -- the result is an fp, Fut[Par[r]].
                  -- Lift the future to a ParT and flatten the Par[Par[t]] to Par[r]
                  join(liftf(fp))
                end))
end
