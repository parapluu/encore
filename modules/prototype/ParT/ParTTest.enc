module ParTTest

import ParT

def toUpperAndSmall(s: String): Par String
  liftv(s.to_upper()) || liftv(s)

def printUpper(s: String): void
  print s.to_upper()

class Test
  def producer(): String
    "producer"

class Main
  def testBindAndFilter(): void
    -- this test filters and creates two versions of the filtered items
    let fut = (new Test).producer()
        p = empty<String>() || liftv("Test") || liftv("Quick") ||
            liftv("Encore") || liftf(fut)
        filteredP = filter<String>(\(s: String) ->
                                     s.contains("k") or s.contains("prod"), p)
        bindP = bind<String, String>(toUpperAndSmall, filteredP)
    in for s in extract(bindP) print(s)

  def testReduce(): void
    -- this test concats items in a ParT
    let p = empty<String>() || liftv("Test") || liftv("Quick") || liftv("Encore")
        concatFn = \(x: String, acc: String) -> x.concatenate(acc)
        red = reduce<String, String>(concatFn, "", p)
        result = get(red)
    in print(result)

  def testForeach(): void
    let p = empty<String>() || liftv("Test") || liftv("Quick") || liftv("Encore")
    in foreach<String>(p, printUpper)

  def testAggregate(): void
    let p = liftv("Test") || liftv("Quick") || liftv("Encore")
        result = aggregate<String, int, int>(p, 0,
                                             \(s: String, acc: int) -> acc + 1,
                                             \(x: int) -> x)
                 >> \(total: int) ->
                      "Aggregation result: ".concatenate(string_from_int(total))
    in print((extract(result))[0])

  def main(): void {
    this.testBindAndFilter();
    this.testReduce();
    this.testForeach();
    this.testAggregate();
  }
