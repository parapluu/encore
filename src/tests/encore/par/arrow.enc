-- This file was automatically converted by encorec

import ParT.Arrow

fun printResult(t : (int, int)) : void
  match t with
    case (k, v) =>
      println("{}: {}", k, v)
    end
  end
end

fun pprint(x : int) : void
  println(x)
end

fun printVoidTuple(t : (int, int)) : (void, void)
  (printResult(t), ())
end

class Main
  def main() : void
    val inc = fun (x : int) => x + 1
    val double = fun (x : int) => x * 2

    -- Test: first :: Arrow a => a b c -> a (b, d) (c, d)
    val ff = first[int, int, int](arr(inc))
    val pr = arr(printResult)
    val ffc = composition(ff, pr)
    runArrow((4, 5), ffc)

    -- Test: second :: Arrow a => a b c -> A (d, b) (d, c)
    val fsecond = second[int, int, int](arr(inc))
    val ffc = composition(fsecond, pr)
    runArrow((4, 5), ffc)

    -- Test: (>>>) :: Control.Category.Category cat => cat a b -> cat b c -> cat a c
    val fn = composition(arr(inc), arr(double))
    val ffc = composition(fn, arr(pprint))
    runArrow(4, ffc)

    -- Test: (&&&) :: Arrow a => a b c -> a b c' -> a b (c, c')
    val fn = branch(arr(inc), arr(double))
    val ffc = composition(fn, arr[(int, int), (void, void)](printVoidTuple))
    runArrow(5, ffc)

    -- Test: (***) :: Arrow a => a b c -> a d e -> a (b, d) (c, e)
    val fn = combine(arr(inc), arr(double))
    val ffc = composition(fn, arr[(int, int), (void, void)](printVoidTuple))
    runArrow((5, 6), ffc)

    -- Test: runParArrow
    --
    -- The ASCII diagram below depics the operations perform on the ParT
    -- for numbers from 0..8
    --
    --
    --          .--- double . inc
    --         /                 \
    --     .- 1                   + ----.
    --    /    \                 /       \
    --   |      .--- double . mod         \
    -- par                                 +  final result
    --   |      .--- double . inc         /|
    --    \    /                 \       / |
    --     .- 2                   + ----.  |
    --         \                 /         |
    --          .--- double . mod         /
    --                                   /
    --        3   ...    ...   ---------.

    val incAndDouble = composition(arr(inc), arr(double))
    val modAndDouble = composition(arr(fun (x: int) => x % 2), arr(double))
    val aggregation = liftA2(fun (x: int, y: int) => x + y, incAndDouble, modAndDouble)
    val par = each([1, 2, 3, 4, 5, 6, 7, 8])
    val pResult = runParArrow(par, aggregation)
    reduce(fun (acc: int, x: int) => acc + x, 0, pResult) ~~> pprint
  end
end
