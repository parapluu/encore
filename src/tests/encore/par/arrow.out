5: 5
4: 6
10
6: 10
6: 12
96